packages/workflow/src/workflow.ts

```diff
@@ -763,39 +762,27 @@ export class Workflow {
 	}
 
 	getParentMainInputNode(node: INode): INode {
-		if (node) {
-			const nodeType = this.nodeTypes.getByNameAndVersion(node.type, node.typeVersion);
-			const outputs = NodeHelpers.getNodeOutputs(this, node, nodeType.description);
+		if (!node) return node;
 
-			if (
-				outputs.find(
-					(output) =>
-						((output as INodeOutputConfiguration)?.type ?? output) !== NodeConnectionTypes.Main,
-				)
-			) {
-				// Get the first node which is connected to a non-main output
-				const nonMainNodesConnected = outputs?.reduce((acc, outputName) => {
-					const parentNodes = this.getChildNodes(
-						node.name,
-						(outputName as INodeOutputConfiguration)?.type ?? outputName,
-					);
-					if (parentNodes.length > 0) {
-						acc.push(...parentNodes);
-					}
-					return acc;
-				}, [] as string[]);
-
-				if (nonMainNodesConnected.length) {
-					const returnNode = this.getNode(nonMainNodesConnected[0]);
-					if (returnNode === null) {
-						// This should theoretically never happen as the node is connected
-						// but who knows and it makes TS happy
-						throw new ApplicationError(`Node "${nonMainNodesConnected[0]}" not found`);
-					}
+		const nodeConnections = this.connectionsBySourceNode[node.name];
+		if (!nodeConnections) return node;
+
+		// Get non-main connection types
+		const nonMainConnectionTypes = Object.keys(nodeConnections).filter(
+			(type) => type !== NodeConnectionTypes.Main,
+		);
 
-					// The chain of non-main nodes is potentially not finished yet so
-					// keep on going
-					return this.getParentMainInputNode(returnNode);
+		for (const connectionType of nonMainConnectionTypes) {
+			const connections = nodeConnections[connectionType] || [];
+			for (const connectionGroup of connections) {
+				for (const connection of connectionGroup || []) {
+					if (connection?.node) {
+						const returnNode = this.getNode(connection.node);
+						if (!returnNode) {
+							throw new ApplicationError(`Node "${connection.node}" not found`);
+						}
+						return this.getParentMainInputNode(returnNode);
+					}
 				}
 			}
 		}
```

Comment:
I prefer the more declarative (functional programming) to the iterative way and if it doesn't decrease performance I'd go with this

```suggestion
		const parentNode = nonMainConnectionTypes
			.flatMap((connectionType) => nodeConnections[connectionType] || [])
			.flatMap((connectionGroup) => connectionGroup || [])
			.map((connection) => {
				if (connection?.node) {
					const returnNode = this.getNode(connection.node);
					if (!returnNode) {
						throw new ApplicationError(`Node "${connection.node}" not found`);
					}
					return this.getParentMainInputNode(returnNode);
				}
			})
			.find(Boolean);

		return parentNode || node;
```